contracts/contracts/strategies/sonic/SonicValidatorDelegator.sol

```diff
@@ -0,0 +1,363 @@
+
+        // Can still withdraw even if the validator is no longer supported
+        // Load the withdrawal from storage into memory
+        WithdrawRequest memory withdrawal = withdrawals[withdrawId];
+        require(!isWithdrawnFromSFC(withdrawId), "Already withdrawn");
+
+        uint256 sBalanceBefore = address(this).balance;
+
+        ISFC(sfc).withdraw(withdrawal.validatorId, withdrawId);
+
+        // Save state to storage
+        withdrawnAmount = address(this).balance - sBalanceBefore;
+        pendingWithdrawals -= withdrawal.undelegatedAmount;
+        withdrawals[withdrawId].undelegatedAmount = 0;
+
+        // Wrap Sonic (S) to Wrapped Sonic (wS)
+        IWrappedSonic(wrappedSonic).deposit{ value: withdrawnAmount }();
+
+        // Transfer the Wrapped Sonic (wS) to the Vault
+        // slither-disable-next-line unchecked-transfer unused-return
+        IERC20(wrappedSonic).transfer(vaultAddress, withdrawnAmount);
+
+        // withdrawal.undelegatedAmount & withdrawnAmount can differ in case of slashing
+        emit Withdrawn(
+            withdrawId,
+            withdrawal.validatorId,
+            withdrawal.undelegatedAmount,
+            withdrawnAmount
+        );
+    }
+
+    // slither-disable-end reentrancy-no-eth
+
+    /// @notice returns a bool whether a withdrawalId has already been withdrawn or not
+    function isWithdrawnFromSFC(uint256 withdrawId) public view returns (bool) {
+        WithdrawRequest memory withdrawal = withdrawals[withdrawId];
+        require(withdrawal.validatorId > 0, "Invalid withdrawId");
+        return withdrawal.undelegatedAmount == 0;
+    }
+
+    /**
+     * @notice Restake any pending validator rewards for all supported validators
+     * @param validatorIds List of Sonic validator IDs to restake rewards
+     */
+    function restakeRewards(uint256[] calldata validatorIds)
+        external
+        nonReentrant
+    {
+        uint256 totalRewards = 0;
+        for (uint256 i = 0; i < validatorIds.length; ++i) {
+            uint256 rewards = ISFC(sfc).pendingRewards(
+                address(this),
+                validatorIds[i]
+            );
+
+            if (rewards > 0) {
+                totalRewards += rewards;
+                ISFC(sfc).restakeRewards(validatorIds[i]);
+            }
+        }
+
+        // The SFC contract will emit Delegated and RestakedRewards events.
+        // The checkBalance function should not change as the pending rewards will moved to the staked amount.
+    }
+
+    /**
+     * @notice To receive native S from SFC and Wrapped Sonic (wS)
+     *
+     * @dev This does not prevent donating S tokens to the contract
+     * as wrappedSonic has a `withdrawTo` function where a third party
+     * owner of wrappedSonic can withdraw to this contract.
+     */
+    receive() external payable {
+        require(
+            msg.sender == sfc || msg.sender == wrappedSonic,
+            "S not from allowed contracts"
```

Comment:
⚪️ This message does not match what the check actually does.

Reply:
changed to `S not from allowed contracts`

Reply:
This function doesn't allow for receiving of S when not from the 2 contracts allowed: sfc & wrappedSonic. What do you think would be a better sounding message?

Reply:
I miss-read. Current message fine.